File: capacity-club-api/src/domain-modules/member-plan-subscription/model/entity/member-plan-subscription.entity.ts
import { UniqueId, uniqueIdTransformer } from '@common/model/unique-id';
import { ApiProperty } from '@nestjs/swagger';
import { Transform } from 'class-transformer';
import { MemberCard } from 'domain-modules/member-card/model/entity/member-card.entity';
import { MemberPlan } from 'domain-modules/member-plan/model/entity/member-plan.entity';
import { Member } from 'domain-modules/member/model/entity/member.entity';
import {
  Entity,
  PrimaryColumn,
  CreateDateColumn,
  ManyToOne,
  JoinColumn,
  OneToOne,
} from 'typeorm';

/**
 * Entity for recording the purchase history of member plans by members.
 */
@Entity()
export class MemberPlanSubscription {
  /**
   * Unique identifier for the member plan subscription.
   * Uses a custom transformer for serialization.
   */
  @PrimaryColumn('varchar')
  @Transform(uniqueIdTransformer.to, { toClassOnly: true })
  @Transform(uniqueIdTransformer.from, { toPlainOnly: true })
  member_plan_subscription_id: UniqueId;

  @CreateDateColumn()
  purchase_date: Date;

  @ManyToOne(() => Member, { lazy: true })
  @JoinColumn({ referencedColumnName: 'member_id', name: 'member_id_fk' })
  @ApiProperty({ type: () => Member })
  member: Promise<Member>;

  @OneToOne(() => MemberPlan, { eager: true })
  @JoinColumn({
    referencedColumnName: 'member_plan_id',
    name: 'member_plan_id_fk',
  })
  member_plan: MemberPlan;

  @OneToOne(() => MemberCard, { eager: true })
  @JoinColumn({
    referencedColumnName: 'member_card_id',
    name: 'member_card_id_fk',
  })
  member_card: MemberCard;
}

File: capacity_club_mobile_app/lib/domain-features/domain/member-plan-subscription/entity/member_plan_subscription_entity.dart
import 'package:capacity_club_mobile_app/core/model/entities/unique_id.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/member-card/entity/member_card_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/member/entity/member_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/member-plan/entity/member_plan_entity.dart';

class MemberPlanSubscriptionEntity {
  final UniqueId member_plan_subscription_id;
  final DateTime purchase_date;
  final MemberEntity member;
  final MemberPlanEntity member_plan;
  final MemberCardEntity member_card;
  final bool is_empty;
  MemberPlanSubscriptionEntity(
      {required this.member_plan_subscription_id,
      required this.purchase_date,
      required this.member,
      required this.member_plan,
      required this.member_card,
      this.is_empty = false});

  factory MemberPlanSubscriptionEntity.empty() {
    return MemberPlanSubscriptionEntity(
        member_plan_subscription_id: UniqueId(''),
        purchase_date: DateTime.now(),
        member: MemberEntity.empty(),
        member_plan: MemberPlanEntity.empty(),
        member_card: MemberCardEntity.empty(),
        is_empty: true);
  }
}
